File: blocksandstuff-blocks/src/main/java/org/everbuild/blocksandstuff/blocks/group/block/VanillaTagBlockGroup.java
package org.everbuild.blocksandstuff.blocks.group.block;

import net.kyori.adventure.key.Key;
import net.minestom.server.MinecraftServer;
import net.minestom.server.gamedata.tags.Tag;
import net.minestom.server.instance.block.Block;
import net.minestom.server.utils.NamespaceID;

import java.util.Collection;
import java.util.Objects;

public class VanillaTagBlockGroup implements BlockGroup {
    private final Key key;

    public VanillaTagBlockGroup(Key key) {
        this.key = key;
    }

    @Override
    public Collection<Block> allMatching() {
        Tag tag = MinecraftServer.getTagManager().getTag(Tag.BasicType.BLOCKS, key.asString());
        assert tag != null;

        return tag
                .getValues()
                .stream()
                .map(NamespaceID::from)
                .map(Block::fromNamespaceId)
                .filter(Objects::nonNull)
                .toList();
    }
}

File: blocksandstuff-blocks/src/main/kotlin/org/everbuild/blocksandstuff/blocks/group/block/VanillaTagBlockGroup.kt
package org.everbuild.blocksandstuff.blocks.group.block

import net.kyori.adventure.key.Key
import net.minestom.server.MinecraftServer
import net.minestom.server.gamedata.tags.Tag
import net.minestom.server.instance.block.Block
import net.minestom.server.utils.NamespaceID

class VanillaTagBlockGroup(private val key: Key) : BlockGroup {
    override fun allMatching(): Collection<Block> {
        val tag = MinecraftServer
            .getTagManager()
            .getTag(Tag.BasicType.BLOCKS, key.asString())!!
        return tag
            .values
            .map(NamespaceID::from)
            .mapNotNull(Block::fromNamespaceId)
            .toList()
    }
}
